src/v/cloud_topics/level_one/metastore/state_update.cc

```diff
@@ -116,6 +119,87 @@ std::expected<std::monostate, stm_update_error> add_objects_update::can_apply(
             }
             expected_next = kafka::next_offset(extent.last_offset);
         }
+        if (!new_terms.contains(tidp)) {
+            return std::unexpected(
+              stm_update_error{fmt::format("No terms requested for {}", tidp)});
+        }
+    }
+    // Now that we've validated the offsets of our extents, validate the terms
+    // for the accepted extents.
+    for (const auto& [tp, req_entries] : new_terms) {
+        if (req_entries.empty()) {
+            return std::unexpected(
+              stm_update_error{
+                fmt::format("Empty terms requested for {}", tp)});
+        }
+        if (corrected_next_offsets.contains(tp)) {
+            continue;
+        }
+        auto extents_it = new_extents.find(tp);
+        if (extents_it == new_extents.end() || extents_it->second.empty()) {
+            return std::unexpected(
+              stm_update_error{fmt::format(
+                "Terms provided for a partition that has no extents", tp)});
+        }
+        auto new_extents_start_offset
+          = extents_it->second.begin()->base_offset();
+        auto new_terms_start_offset = req_entries.begin()->start_offset;
+        if (new_extents_start_offset != new_terms_start_offset) {
```

Comment:
Do we also check that all the start offsets for terms are also before the last offset in objects?

Reply:
Basically we need a term that starts with our offsets, and all term offsets are increasing, and all term offsets are <= last_offset

Reply:
Done